WIP: debug report & vkCreateInstance wording update

courtney-g opened the pull request:
Fixes #490, #489, #483, #478

krOoze commented on doc/specs/vulkan/chapters/initialization.txt:
```diff
@@ -172,14 +172,21 @@ include::../api/protos/vkCreateInstance.txt[]
+For example, if an application wants to enable extensionB that is only
+supported in layerB, the application needs to include "layerB" in
+sname:ppEnabledLayerNames and "extensionB" in the
+sname:ppEnabledExtensionNames list.
```
- This whole thing above renders as a single large monolithic paragraph -> unesthetical and hard to read.
 - If I understand things, the layer supported extension case description is now redundant to the descriptions above. It could be a Note block instead (would also help with the big paragraph problem)

courtney-g commented:
> Revision number in appendix is not incremented.

Not sure which revision number you are referring to. I updated the one in vk.xml for the debug_report extension.

> there possibly should be VU that VkDebugReportCallbackCreateInfoEXT::pfnCallback must be a pointer to a valid function (see VkAllocationCallbacks for examples). Should IMO be generated as Implicit VU instead though,

Probably should be implicit. That requires generator script updates. I'll put in manual VU for now.

> Missing pAllocator descriptions for create and destroy callback command
got it

> The VUs for pAllocator seem outdated in debug_report (compared e.g. to the VkInstance versions)

The VU is same as everywhere else it's auto generated. If you don't think that's appropriate, please file issue on it.

> Possibly copy-paste error for many vkDebugReportMessageEXT parameter descriptions (e.g. "messageCode is a layer-defined value indicating what test triggered this callback." — I would assume it is somewhat app-defined in this case as is the test)

I'll change location and messageCode to be application defined values for this call.

> The synchronization here (due to the nature of this extension) seem non-obvious. E.g. can I call vkDestroyDebugReportCallbackEXT from the callback function? Or on different thread while some vkX command (which is subject that can cause a callback) has not yet returned?

In the description of vkDestroyDebugReportCallbackEXT the 'callback' parameter is defined to be "externally synchronized" which is:
External synchronization
  A type of synchronization required of the application, where parameters defined to be externally synchronized must not be used simultaneously in multiple threads.

For this call, that may not be enough. Even if the application were to ensure that only one thread is calling vkDestroyDebugReportCallbackEXT making such a call during the callback could result in undefined behavior. Generally modifying a list while iterating through the list is not a good thing to do. I'll look at adding a VU for that.

> It is sneakily implied by the "bitwise-AND" description, but it would potentially be nice to say in PFN_vkDebugReportCallbackEXT::flags that it won't ever be 0

Why? Not obvious to me how adding such a statement is "potentially nice". To me it's another must: to be validated.

> vkCreateDevice should have similar description as vkCreateInstance now has.

updated

could also fix random minor issues in debug_report

> Many undecorated or wrong usage of "may" across the debug_report extension text.

You'll need to point those out. A lot of the "may"'s refer to example / speculative behavior.

> Undecorated and misnamed DebugReportCallback in pUserData parameter desc. of PFN_vkDebugReportCallbackEXT

I think I got it.

> there is "This extension define" instead "defines" in appendix.

Got it

> undecorated "RenderPass load_op"

I think I got it.

Rest as line comments.

krOoze commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -79,14 +66,18 @@ include::../api/enums/VkDebugReportFlagBitsEXT.txt[]
+For each non-zero result the corresponding callback will be called.
 The callback will come directly from the component that detected the event,
 unless some other layer intercepts the calls for its own purposes (filter
 them in different way, log to system error log, etc.) An application may
```
@courtney-g That's perhaps for internal discussion.

BUT IMO RFC/Terminology chapter does not define any difference between decorated and non-decorated word &mdash; it simply says "The key words [keywords] in this document are to be interpreted as [...]."

Therfore (IMO) failing to decorate the keywords are just an violation of chosen style of this spec.
And using them colloquially (whether decorated or not) is violation of the Terminology chapter.

krOoze commented:
> Not sure which revision number you are referring to.

Right, sorry. I mean this one specifically (and also the "last modified date" there too):
https://github.com/KhronosGroup/Vulkan-Docs/blob/170410-debug_report_issues/doc/specs/vulkan/appendices/VK_EXT_debug_report.txt#L13


> > The VUs for pAllocator seem outdated in debug_report (compared e.g. to the VkInstance versions)
> 
> The VU is same as everywhere else it's auto generated. If you don't think that's appropriate, please file issue on it.

Sorry, I must have misread something.

Nevertheless there seem to be a copy-paste error in `vkDestroyDebugReportCallbackEXT()` VU for it.
"If VkAllocationCallbacks were provided when **instance** was created [...]"
Should be `callback` instead od `instance`? Elsewhere in the spec it seems to always be the object being vkDestroyed (even for non-dispatchables like `VkFence`)

> > It is sneakily implied by the "bitwise-AND" description, but it would potentially be nice to say in PFN_vkDebugReportCallbackEXT::flags that it won't ever be 0
>
> Why? Not obvious to me how adding such a statement is "potentially nice". To me it's another must: to be validated.

Well, it just seemed like a surprising implication of the `&` (events with `0` flag won't pass through no matter the filter flag). Also someone reading just the `PFN_vkDebugReportCallbackEXT` definition, which comes later, might miss that info &mdash; other parameters seem to have described there what they can and cannot be (except perhaps `pLayerPrefix` which should say it's null-terminated string, to disallow `nullptr` there)

It should not require another validation. As an Vulkan output it is not a subject for layer validation.

> You'll need to point those out. A lot of the "may"'s refer to example / speculative behavior.

That's perhaps for your internal discussion with the Group.

I mean all of them undecorated ones.

IMO RFC/Terminology chapter does not define any difference between decorated and non-decorated word &mdash; it simply says "The key **words** [list of keywords] in this document are to be interpreted as [...]."

Therfore my thinking is:
Failing to decorate any keyword is just a violation of the chosen style guide for this spec.
And using them colloquially (whether decorated or not) is a violation of the Terminology chapter.

krOoze commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -223,14 +231,11 @@ include::../api/protos/vkDebugReportMessageEXT.txt[]
   * pname:objectType is a elink:VkDebugReportObjectTypeEXT specifying the
```
These still feel copy-pasted and not applicable to `vkDebugReportMessageEXT`.
Should be something like:

> * `instance` is the debug stream's `VkInstance`.
> * `flags` indicates the `VkDebugReportFlagBitsEXT` classification of this event/message.

krOoze commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -223,14 +231,11 @@ include::../api/protos/vkDebugReportMessageEXT.txt[]
+  * pname:location is an application defined value.
+  * pname:messageCode is an application defined value.
   * pname:pLayerPrefix is the abbreviation of the component making the
     callback.
```
Same problem. Should be something like "making this event/message" instead "making the callback".

krOoze commented on doc/specs/vulkan/chapters/initialization.txt:
```diff
@@ -172,14 +172,16 @@ include::../api/protos/vkCreateInstance.txt[]
+After verifying and enabling the instance layers and extensions the
+sname:VkInstance object is created and returned to the application.
+If a requested extension is only supported by a layer, both the layer and
+extension need to be specified at fname:vkCreateInstance time.
```
Would add " , for the creation to succeed." at the end to emphasize this is not a requirement. (Same for the `vkCreateDevice` copy)

Should be "and **the** extension"?

krOoze commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -124,11 +128,15 @@ include::../api/funcpointers/PFN_vkDebugReportCallbackEXT.txt[]
   * pname:pMessage is a null-terminated string detailing the trigger
```
Should perhaps also say pname:pLayerPrefix is a null-terminated string (i.e. it cannot be e.g. `nullptr` I assume)

courtney-g commented:
FYI - pushed updated CLs

krOoze commented on doc/specs/vulkan/chapters/devsandqueues.txt:
```diff
@@ -560,6 +560,14 @@ include::../api/protos/vkCreateDevice.txt[]
+After verifying and enabling the extensions the sname:VkDevice object is created
+and returned to the application.
+If a requested extension is only supported by a layer, the
+layer needs to be specified at flink:vkCreateInstance time.
```
should probably match the `vkCreateInstance` wording, i.e. the changed:
"(e.g. in the implementation or in any enabled instance layer)"
and
"[...] for the creation to succeed."

courtney-g replied:
ok

krOoze commented on doc/specs/vulkan/chapters/initialization.txt:
```diff
@@ -172,14 +172,17 @@ include::../api/protos/vkCreateInstance.txt[]
+After verifying and enabling the instance layers and extensions the
+sname:VkInstance object is created and returned to the application.
+If a requested extension is only supported by a layer, both the layer and
+the extension need to be specified at fname:vkCreateInstance time, for the
```
Not an englishman, but shouldn't this be without `,` (because no verb)?

courtney-g replied:
ok

courtney-g commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -217,22 +223,19 @@ To inject its own messages into the debug stream, call:
     triggered.
-  * pname:object is object where the issue was detected.
+  * pname:object object where the issue was detected.
     pname:object may be ename:VK_NULL_HANDLE if there is no object
```
Yeah, that's an appropriate spot for can:

courtney-g commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -217,22 +223,19 @@ To inject its own messages into the debug stream, call:
+    event/message.
   * pname:pMessage is a null-terminated string detailing the trigger
     conditions.
 
```
Right, still apply flags.

courtney-g commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -257,12 +260,18 @@ include::../api/protos/vkDestroyDebugReportCallbackEXT.txt[]
  * pname:callback the sname:VkDebugReportCallbackEXT object to destroy.
+   pname:callback is an externally synchronized object and cannot be used
+   on more than one thread at a time. This means that
+   fname:vkDestroyDebugReportCallbackEXT cannot be called when a callback
```
ok

courtney-g commented:
CLs updated.

krOoze commented on doc/specs/vulkan/chapters/devsandqueues.txt:
```diff
@@ -560,6 +560,15 @@ include::../api/protos/vkCreateDevice.txt[]
     returned.
 
+fname:vkCreateDevice verifies that the requested extensions are supported (e.g.
+the in implementation or in any enabled instance layer) and if any requested extension is
```
typo: "the in"

krOoze commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -111,24 +113,28 @@ include::../api/funcpointers/PFN_vkDebugReportCallbackEXT.txt[]
     triggered.
   * pname:object gives the object where the issue was detected.
-    pname:object may be ename:VK_NULL_HANDLE if there is no object
+    pname:object can: be ename:VK_NULL_HANDLE if there is no object
```
Wrong spot. This one should probably be "may:" (because it is basically output from Vulkan).

krOoze commented on doc/specs/vulkan/chapters/VK_EXT_debug_report.txt:
```diff
@@ -217,27 +223,24 @@ To inject its own messages into the debug stream, call:
     triggered.
-  * pname:object is object where the issue was detected.
+  * pname:object object where the issue was detected.
     pname:object may be ename:VK_NULL_HANDLE if there is no object
```
I meant "can:" here.

TobiasHector commented:
@courtney-g it looks like this got fixed internally a few months back with what appears to be similar, if not the same, changes. Not sure what happened, but think we can close this? Can you confirm and close if true?
